### PR TITLE
feat: Use pydantic instead of weave.type

### PR DESCRIPTION
**Pull request by shawnlewis**
This makes it possible to replace current uses of the dataclasses based @weave.type() with pydantic models.

- new Object/Dataset/Model in the flow directory. They are weave objects built with pydantic.
- fixes all object types to have a .name attribute (before that would collide with Type's .name property)
- weave.publish will use the .name attribute on an object as the published object/artifact name, if it exists. The "name" attribute conflicts with the weave_types.Type.name property, so I had to put fixes in a bunch of places to work around the collision.
- non-builtin ops are no longer registered by default, which means they are not written to disk by default as a local artifact when they are declared, which makes boot-up faster! A lot of this PR is updating tests because of this change

This does not yet switch the "weave.weaveflow" objects over to use the new Pydantic style. There is a new directory "flow" that has Object, Model, Dataset. Will update examples and remove weave.weaveflow in a separate PR.

### CODE REVIEW

File: weave/weaveflow/obj.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        ):
+            bound_op_def = typing.cast(BoundOpDef, bound_op_def)
+            unbound_op_def = bound_op_def.op_def
+            if unbound_op_def.name.startswith("op-"):
+                unbound_op_def.name = f"{name}-{unbound_op_def.name[3:]}"
+        return original_class
+
+
+class Object(pydantic.BaseModel, metaclass=ObjectMeta):
+    name: Optional[str] = "hello"
```
Comment (tssweeney) [inline weave/weaveflow/obj.py:25]: probably don't want to default the name to "hello"

File: weave/flow/dataset.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import typing
+import weave
+
+from .obj import Object
```
Comment (tssweeney) [inline weave/flow/dataset.py:4]: here the import is relative, elsewhere it is `from weave.flow.obj import Object` would be good to be consistent

File: weave/mappers_python_def.py [hunk 3]
```diff
@@ -66,9 +68,10 @@ def apply(self, obj):
         constructor_sig = inspect.signature(instance_class)
         for k, serializer in self._property_serializers.items():
             if serializer.type != OpDefType() and k in constructor_sig.parameters:
-                # None haxxx
-                # TODO: remove
-                obj_val = obj.get(k)
+                get_k = k
```
Comment (tssweeney) [inline weave/mappers_python_def.py:71]: would be nice to comment on this special case

Comment (shawnlewis) [reply weave/mappers_python_def.py]: This hacking name stuff has all been centralized into the types layer and commented.

File: weave/flow/dataset.py (hunk 2 above)
Comment (shawnlewis) [reply weave/flow/dataset.py]: fixed.

**Review by tssweeney — APPROVED** [review]

**Pull request merged**